design-decisions/2020-05-02-function-rules.md

```diff
@@ -0,0 +1,261 @@
+
+## Solution
+
+The interpretation of simplification rules will not change,
+but the family of K rules defining `fun`,
+
+```.k
+rule fun(φ₁ᵢ({Y}), φ₂ᵢ({Y}), ...) => ψᵢ({Y}) requires Preᵢ({Y}) ensures Postᵢ({Y})
+```
+
+will be interpreted in Kore as
+
+```.kore
+axiom \implies(Preᵢ({Y}) ∧ Argsᵢ({X}, {Y}) ∧ Prioᵢ({X}), (fun(X₁, X₂, ...) = ψᵢ({Y})) ∧ Postᵢ({Y}))
+```
+
+where
+
+```.kore
+Argsᵢ({X}, {Y}) = (X₁ ∈ φ₁ᵢ({Y})) ∧ (X₂ ∈ φ₂ᵢ({Y})) ∧ ...
+Prioᵢ({X}) = ∧ⱼ ¬ ∃ {Y}. Preⱼ({Y}) ∧ Argsⱼ({X}, {Y})
+    for all j that priority(rule j) < priority(rule i).
+```
+
+The predicate `Argsᵢ` interprets the argument patterns `φ₁ᵢ` element-wise,
+matching the user's intent and (as we will see below) avoiding the problems described above.
+The predicate `Prioᵢ` encodes the `priority` attribute in the pre-condition of the rule,
+which is now possible only because we have moved argument matching into the `Argsᵢ` pre-condition.
+
+### Partial heads
+
+The troublesome example,
+
+```.k
+rule sizeMap(_ |-> _ M:Map) = 1 +Int sizeMap(M)
+```
+
+is interpreted in Kore as
+
+```.kore
+axiom \implies(X ∈ concatMap(elementMap(_1, _2), M:Map), sizeMap(X) = +Int(1, sizeMap(M))).
+```
+
+If `concatMap(_, _)` is undefined,
+then the pre-condition `X ∈ concatMap(_, _)` is not satisfied;
+therefore, the rule is sound.
+
+### Or patterns
+
+The example or-pattern rule,
+
+```.k
+rule fun(A #Or B) => C
+```
+
+will be interpreted as
+
+```.kore
+axiom \implies(X ∈ (A ∨ B), fun(X) = C).
+```
+
+The disjunction distributes over `_ ∈ _`
+
+```.kore
+X ∈ (A ∨ B)
+= (X ∈ A) ∨ (X ∈ B)
+```
+
+and over `\implies`
+
+```.kore
+\implies(X ∈ (A ∨ B), fun(X) = C)
+= \implies(X ∈ A, fun(X) = C) ∧ \implies(X ∈ B, fun(X) = C)
+```
+
+so that the original axiom is equivalent to two axioms, as intended:
+
+```.kore
+axiom \implies(X ∈ A, fun(X) = C)
+axiom \implies(X ∈ B, fun(X) = C).
+```
+
+### Priority
+
+The `priority` and `owise` attributes are now encoded explicitly in the
+pre-condition of the axiom.
+Consider a function,
+
+```.k
+rule L:Int <= X:Int < _     => false requires notBool L <=Int X
+rule _     <= X:Int < U:Int => false requires notBool X  <Int U [priority(51)]
+rule _     <= _     < _     => true                             [owise]
+```
+
+which will be interpreted in Kore as axioms
+
+```.kore
+axiom
+    \implies(
+        \and(
+            true = notBool(<=Int(L, X)),
+            (X₁ ∈ L) ∧ (X₂ ∈ X)
+        ),
+        \and(
+            _<=_<_(X₁, X₂, X₃) = false,
+            \top
+        )
+    )
+axiom
+    \implies(
+        \and(
+            true = notBool(<Int(X, U)),
+            (X₂ ∈ X) ∧ (X₃ ∈ U),
+            \not \exists L X.
+                \and(
+                    true = notBool(<=Int(L, X)),
+                    (X₁ ∈ L) ∧ (X₂ ∈ X)
+                )
+        ),
+        \and(
+            _<=_<_(X₁, X₂, X₃) = false,
+            \top
+        )
+    )
+axiom
+    \implies(
+        \and(
+            \top,
+            \top,
```

Comment:
I'm reading this as `\and(\top, \top, ...)`. Should one of the `\top` be removed?

Reply:
No, there will be three pre-conditions now: the actual `requires`, the `Args`, and the `Prio`. Only the `Prio` can be omitted if the rule has the lowest priority (that is, if there are no other rules to apply first).